### PR TITLE
feat: implement some expr side effect detector

### PR DESCRIPTION
**Pull request by woai3c**
### Description
I have enhanced some of the code for side-effect detection in expressions, but there are still parts that I'm currently unsure how to implement.

### CODE REVIEW

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 1]
```diff
@@ -161,23 +164,29 @@ impl<'a> SideEffectDetector<'a> {
...
-
+      Expression::BinaryExpression(binary_expr) => {
+        // For binary expressions, both sides could potentially have side effects
+        self.detect_side_effect_of_expr(&binary_expr.left)
+          || self.detect_side_effect_of_expr(&binary_expr.right)
+      }
+      Expression::PrivateInExpression(private_in_expr) => {
+        self.detect_side_effect_of_expr(&private_in_expr.right)
+      }
+      Expression::MetaProperty(_) | Expression::ThisExpression(_) => false,
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]: I think `this.xxx` is considering may have side effect.

We could consider`import.meta` doesn't have side effect, but `import.meta.xxx` is tricky and I would like to consider it have side effect so far.

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 2]
```diff
@@ -161,23 +164,29 @@ impl<'a> SideEffectDetector<'a> {
...
+      Expression::AwaitExpression(await_expr) => {
+        self.detect_side_effect_of_expr(&await_expr.argument)
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]: Consider case like
```js
let thenable = {
  then() {
     console.log('hello')
   }
}

await thenable
```

**Review by hyf0 — CHANGES_REQUESTED** [review]

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 3]
```diff
@@ -526,4 +536,47 @@ mod test {
...
+    assert!(!get_statements_side_effect("const a = 1; const b = 2; a + b"));
+    // accessing global variable may have side effect
+    assert!(get_statements_side_effect("1 + foo"));
+    assert!(get_statements_side_effect("2 + bar"));
+  }
+
+  #[test]
+  fn test_private_in_expression() {
+    assert!(!get_statements_side_effect("#privateField in this"));
+    assert!(!get_statements_side_effect("#privateField in this.x"));
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]:
```suggestion
    assert!(get_statements_side_effect("#privateField in this.x"));
```

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 4]
```diff
@@ -526,4 +536,47 @@ mod test {
...
+    // accessing global variable may have side effect
+    assert!(get_statements_side_effect("#privateField in bar"));
+    assert!(get_statements_side_effect("#privateField in foo"));
+  }
+
+  #[test]
+  fn test_this_expression() {
+    assert!(!get_statements_side_effect("this"));
+    assert!(!get_statements_side_effect("this.a"));
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]:
```suggestion
    assert!(get_statements_side_effect("this.a"));
```

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 5]
```diff
@@ -526,4 +536,47 @@ mod test {
...
+    assert!(!get_statements_side_effect("this.a + this.b"));
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]:
```suggestion
    assert!(get_statements_side_effect("this.a + this.b"));
```

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 6]
```diff
@@ -526,4 +536,47 @@ mod test {
...
+    assert!(get_statements_side_effect("this.a = 10"));
+  }
+
+  #[test]
+  fn test_meta_property_expression() {
+    assert!(!get_statements_side_effect("import.meta"));
+    assert!(!get_statements_side_effect("import.meta.url"));
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]:
```suggestion
    assert!(get_statements_side_effect("import.meta.url"));
```

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs (hunk 1 above)
Comment (woai3c) [reply crates/rolldown/src/ast_scanner/side_effect_detector.rs]: ok

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 7]
```diff
@@ -526,4 +536,47 @@ mod test {
...
+    assert!(!get_statements_side_effect("{ url } = import.meta"));
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]:
```suggestion
    assert!(get_statements_side_effect("{ url } = import.meta"));
```

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs (hunk 2 above)
Comment (woai3c) [reply crates/rolldown/src/ast_scanner/side_effect_detector.rs]: ok

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs [hunk 8]
```diff
@@ -526,4 +536,47 @@ mod test {
...
+  }
+
+  #[test]
+  fn test_await_expression() {
+    assert!(!get_statements_side_effect("await 1 + 1"));
+    assert!(!get_statements_side_effect("const a = 1; const b = 2; await a + b"));
+    assert!(get_statements_side_effect("await foo + bar"));
+    assert!(get_statements_side_effect("await foo"));
+    assert!(get_statements_side_effect("await fetch('https://example.com')"));
+  }
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/side_effect_detector.rs]:
```suggestion
     assert!(get_statements_side_effect("let foo; await foo;"));
  }
```

**Review by hyf0 — CHANGES_REQUESTED** [review]

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs (hunk 2 above)
Comment (woai3c) [reply crates/rolldown/src/ast_scanner/side_effect_detector.rs]: The modification code for 'AwaitExpression' has been removed.

Comment (woai3c) [reply crates/rolldown/src/ast_scanner/side_effect_detector.rs]: > Consider case like
> 
> ```js
> let thenable = {
>   then() {
>      console.log('hello')
>    }
> }
> 
> await thenable
> ```

 The modification code for 'AwaitExpression' has been removed.

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs (hunk 1 above)
Comment (woai3c) [reply crates/rolldown/src/ast_scanner/side_effect_detector.rs]: > I think `this.xxx` is considering may have side effect.
> 
> We could consider`import.meta` doesn't have side effect, but `import.meta.xxx` is tricky and I would like to consider it have side effect so far.

I have made modifications to the code based on your feedback. Please take a look again to see if there are any issues with the code.

File: crates/rolldown/src/ast_scanner/side_effect_detector.rs (hunk 2 above)
Comment (woai3c) [reply crates/rolldown/src/ast_scanner/side_effect_detector.rs]: > Consider case like
> 
> ```js
> let thenable = {
>   then() {
>      console.log('hello')
>    }
> }
> 
> await thenable
> ```

this case always be false.

Comment (hyf0) [reply crates/rolldown/src/ast_scanner/side_effect_detector.rs]: Yeah. But it has side effect, so it should be true.

**Comment by woai3c** [comment]
@hyf0  The code for `AwaitExpression` has been reset. Both `ThisExpression` and `MetaProperty` have been modified as per your suggestions. Please review the updated code. Thank you.

By the way, I'd like to contribute more to Rolldown. Currently, I usually look for TODO comments in the code and then implement them. However, some TODOs seem quite challenging to implement. Could you give me some advice on what aspects of Rolldown I should focus on to contribute more effectively?

**Review by hyf0 — APPROVED** [review]
Thank you!

**Pull request merged**